mailbox/cassandra/src/test/java/org/apache/james/mailbox/cassandra/mail/CassandraDeletedMessageDAOTest.java

```diff
@@ -0,0 +1,225 @@
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+public class CassandraDeletedMessageDAOTest {
+    public static final CassandraId MAILBOX_ID = CassandraId.timeBased();
+    public static final MessageUid UID_1 = MessageUid.of(1);
+    public static final MessageUid UID_2 = MessageUid.of(2);
+    public static final MessageUid UID_3 = MessageUid.of(3);
+    public static final MessageUid UID_4 = MessageUid.of(4);
+    public static final MessageUid UID_7 = MessageUid.of(7);
+    public static final MessageUid UID_8 = MessageUid.of(8);
+
+    private CassandraCluster cassandra;
+    private CassandraDeletedMessageDAO testee;
+
+    @Before
+    public void setUp() {
+        cassandra = CassandraCluster.create(
+            new CassandraDeletedMessageModule());
+        cassandra.ensureAllTables();
+
+        testee = new CassandraDeletedMessageDAO(cassandra.getConf());
+    }
+
+    @After
+    public void tearDown() {
+        cassandra.clearAllTables();
+    }
+
+    @Test
+    public void retrieveDeletedMessageShouldReturnEmptyByDefault() {
+        List<MessageUid> result = testee
+            .retrieveDeletedMessage(MAILBOX_ID, MessageRange.all())
+            .join()
+            .collect(Guavate.toImmutableList());
+
+        assertThat(result).isEmpty();
+    }
+
+    @Test
+    public void addDeletedMessageShouldThenBeReportedAsDeletedMessage() {
+        testee.addDeleted(MAILBOX_ID, UID_1).join();
+        testee.addDeleted(MAILBOX_ID, UID_2).join();
+
+        List<MessageUid> result = testee.retrieveDeletedMessage(MAILBOX_ID, MessageRange.all())
+            .join()
+            .collect(Guavate.toImmutableList());
+
+        assertThat(result).containsExactly(UID_1, UID_2);
+    }
+
+    @Test
+    public void addDeletedMessageShouldBeIdempotent() {
+        testee.addDeleted(MAILBOX_ID, UID_1).join();
+        testee.addDeleted(MAILBOX_ID, UID_1).join();
+
+        List<MessageUid> result = testee.retrieveDeletedMessage(MAILBOX_ID, MessageRange.all())
+            .join()
+            .collect(Guavate.toImmutableList());
+
+        assertThat(result).containsExactly(UID_1);
+    }
+
+
+    @Test
+    public void removeUnreadShouldReturnWhenNoData() {
+        testee.removeDeleted(MAILBOX_ID, UID_1).join();
+
+        List<MessageUid> result = testee
+            .retrieveDeletedMessage(MAILBOX_ID, MessageRange.all())
+            .join()
+            .collect(Guavate.toImmutableList());
+
+        assertThat(result).isEmpty();
+    }
+
+    @Test
+    public void removeDeletedMessageShouldNotAffectOtherMessage() {
+        testee.addDeleted(MAILBOX_ID, UID_2).join();
+        testee.addDeleted(MAILBOX_ID, UID_1).join();
+
+        testee.removeDeleted(MAILBOX_ID, UID_1).join();
+
+        List<MessageUid> result = testee
+            .retrieveDeletedMessage(MAILBOX_ID, MessageRange.all())
+            .join()
+            .collect(Guavate.toImmutableList());
+
+        assertThat(result).containsExactly(UID_2);
```

Comment:
Do not need assert in the middle of testing method. We have another test to cover it so just Given/When/Then